metadata-ingestion-modules/airflow-plugin/src/datahub_airflow_plugin/datahub_listener.py

```diff
@@ -0,0 +1,411 @@
+            logger.debug("Merging start datajob into finish datajob")
+            datajob.inlets.extend(original_datajob.inlets)
+            datajob.outlets.extend(original_datajob.outlets)
+            datajob.fine_grained_lineages.extend(original_datajob.fine_grained_lineages)
+
+            for k, v in original_datajob.properties.items():
+                if k not in datajob.properties:
+                    datajob.properties[k] = v
+
+        # TODO: Deduplicate inlets/outlets.
+
+        # Write all other OL facets as DataHub properties.
+        if task_metadata:
+            for k, v in task_metadata.job_facets.items():
+                datajob.properties[f"openlineage_job_facet_{k}"] = Serde.to_json(
+                    redact_with_exclusions(v)
+                )
+
+            for k, v in task_metadata.run_facets.items():
+                datajob.properties[f"openlineage_run_facet_{k}"] = Serde.to_json(
+                    redact_with_exclusions(v)
+                )
+
+    @hookimpl
+    @run_in_thread
+    def on_task_instance_running(
+        self,
+        previous_state,
+        task_instance: "TaskInstance",
+        session: "Session",  # This will always be QUEUED
+    ):
+        self._set_log_level()
+
+        # This if statement mirrors the logic in https://github.com/OpenLineage/OpenLineage/pull/508.
+        if not hasattr(task_instance, "task"):
+            logger.warning(
+                f"No task set for task_id: {task_instance.task_id} - "
+                f"dag_id: {task_instance.dag_id} - run_id {task_instance.run_id}"
+            )
+            return
+
+        logger.debug(
+            f"DataHub listener got notification about task instance start for {task_instance.task_id}"
+        )
+
+        # Render templates in a copy of the task instance.
+        # This is necessary to get the correct operator args in the extractors.
+        task_instance = copy.deepcopy(task_instance)
+        task_instance.render_templates()
+
+        dagrun: "DagRun" = task_instance.dag_run
+        task = task_instance.task
+        dag = task.dag
+
+        self._task_holder.set_task(task_instance)
+
+        # Handle async operators in Airflow 2.3 by skipping deferred state.
+        # Inspired by https://github.com/OpenLineage/OpenLineage/pull/1601.
+        if task_instance.next_method is not None:
+            return
+
+        datajob = AirflowGenerator.generate_datajob(
+            cluster=self.config.cluster,
+            task=task,
+            dag=dag,
+            capture_tags=self.config.capture_tags_info,
+            capture_owner=self.config.capture_ownership_info,
+        )
+
+        # TODO: Make use of get_task_location to extract github urls.
+
+        # Add lineage info.
+        self._extract_lineage(datajob, dagrun, task)
+
+        # TODO: Add handling for Airflow mapped tasks using task_instance.map_index
+
+        datajob.emit(self.emitter, callback=self._make_emit_callback())
```

Comment:
Doesn't matter for this PR, but I'd prefer something like `self.emitter.emit_entity(datajob, callback)` to more closely match our mcp emit call

Reply:
yup I agree - I'm not super happy with the whole datajob/dataflow interfaces